qutipy/general_functions/__init__.py

```diff
@@ -1,16 +1,487 @@
+    """
+
+    return X.conj().T
+
+
+def eye(n):
+    """
+    Generates the n x n identity matrix.
+    """
+
+    return np.identity(n, dtype=float)
+
+
+def generate_all_kets(dims):
+    """
+    Generates the tensor-product orthonormal basis corresponding to vector spaces
+    with dimensions in the list dims.
+
+    ------------------------
+    Example:
+
+        generate_all_kets([2,3]) returns a list containing
+
+            |0,0>
+            |0,1>
+            |0,2>
+            |1,0>
+            |1,1>
+            |1,2>
+
+    """
+
+    dims_set = [range(d) for d in dims]
+
+    L = list(itertools.product(*dims_set))
+
+    K = []
+
+    for l in L:
+        K.append(ket(dims, l))
+
+    return K
+
+
+def get_subblock(X, sys, indices, dim):
+    """For the multipartite operator X, with local dimensions given by the list dim,
+    this function extracts the subblock such that the systems in sys have the
+    values given by those in indices. indices should be a list of tuples, each
+    tuple corresponding to a system in sys. For each tuple, the first element
+    gives the row and the second the column.
+
+    For example, the subblock (<i| ⊗ id ⊗ <j|)X(|k> ⊗ id ⊗ |l>)
+    is given by get_subblock(X,[1,3],[(i,k),(j,l)],[dim1,dim2,dim3]), so the result
+    is a dim2 x dim2 matrix, which is the desired subblock.
+
+
+    Args:
+        X (_type_): _description_
+        sys (_type_): _description_
+        indices (_type_): _description_
+        dim (_type_): _description_. By default it will tak the dimensions from X.
+
+    Returns:
+        _type_: _description_
+    """
+
+    X = np.array(X)
+
+    X_reshape = np.reshape(X, dim + dim)
+
+    num_sys = len(dim)
+    total_sys = range(1, num_sys + 1)
+    dims_keep = [dim[s - 1] for s in list(set(total_sys) - set(sys))]
+
+    to_slice = [slice(None) for i in range(2 * num_sys)]
+
+    count = 0
+    for s in total_sys:
+        if s in sys:
+            to_slice[s - 1] = indices[count][0]
+            to_slice[num_sys + s - 1] = indices[count][1]
+            count += 1
+
+    X_reshape = X_reshape[tuple(to_slice)]
+
+    X_reshape = np.reshape(X_reshape, dims_keep + dims_keep)
+
+    return X_reshape
+
+
+def ket(dim, *args):
+    """
+    Generates a standard basis vector in dimension dim.
+
+    For example, ket(2,0)=|0> and ket(2,1)=|1>.
+
+    In general, ket(d,j), for j between 0 and d-1, generates a column vector
+    (as a numpy matrix) in which the jth element is equal to 1 and the rest
+    are equal to zero.
+
+    ket(d,[j1,j2,...,jn]) generates the tensor product |j1>|j2>...|jn> of
+    d-dimensional basis vectors.
+
+    If dim is specified as a list, then, e.g., ket([d1,d2],[j1,j2]) generates the
+    tensor product |j1>|j2>, with the first tensor factor being d1-dimensional
+    and the second tensor factor being d2-dimensional.
+    """
+
+    # TODO: Dimensions are probably not required
```

Comment:
In general, the dimensions are required, because it affects the size of the array at the output.

Reply:
Fixed. Refer to https://github.com/arnavdas88/QuTIpy/blob/master/qutipy/general_functions.py#L125